pytensor/tensor/subtensor.py

```diff
@@ -145,15 +151,34 @@ def as_index_constant(
         )
     elif isinstance(a, int | np.integer):
         return ps.ScalarConstant(ps.int64, a)
-    elif not isinstance(a, Variable):
-        return as_tensor_variable(a)
-    else:
+    elif isinstance(a, Variable):
         return a
+    return as_tensor_variable(a)
+
+
+@overload
+def as_index_literal(idx: int | np.integer) -> int | np.integer: ...
+
+
+@overload
+def as_index_literal(idx: None) -> None: ...
+
+
+@overload
+def as_index_literal(idx: slice) -> slice: ...
+
+
+@overload
+def as_index_literal(idx: Constant) -> int | np.integer | None: ...
```

Comment:
You can be more specific by using `NoneConst`, `SliceConst`, `Tensor|ScalarConstant` and `SliceVariable`, `Tensor|ScalarVariable` in the signatures

Reply:
TensorContants are already a subclass of TensorVariable (so you don't need that distinction at the signature level), but the other types `ScalarConstant | SliceConstant` may not be. You can make them if you want.

Reply:
This is the PR where we made some of those subclasses: https://github.com/pymc-devs/pytensor/pull/628